packages/core/src/render3/di.ts

```diff
@@ -514,7 +514,10 @@ export function getNodeInjectable(
       value = lData[index] = factory.factory(null, tData, lData, tNode);
       const tView = lData[TVIEW];
       if (value && factory.isProvider && value.ngOnDestroy) {
-        (tView.destroyHooks || (tView.destroyHooks = [])).push(index, value.ngOnDestroy);
+        const type = value.constructor as Type<any>;
+        if (!getComponentDef(type) && !getDirectiveDef(type)) {
+          (tView.destroyHooks || (tView.destroyHooks = [])).push(index, value.ngOnDestroy);
```

Comment:
I know this was existing logic, but I just noticed that we are pushing to `tView.destroyHooks` on every pass, not just the first one. We should wrap that logic in a guard to prevent duplicate pushes:

```suggestion
if (tView.firstTemplatePass) {
   (tView.destroyHooks || (tView.destroyHooks = [])).push(index, value.ngOnDestroy);
}
```

Reply:
Update: looks like it will be a more complicated fix because of things like `injector.get`, so we are punting as out of scope for this PR